File: rust_src/121.买卖股票的最佳时机.rs
/*
 * @lc app=leetcode.cn id=121 lang=rust
 *
 * [121] 买卖股票的最佳时机
 */

// @lc code=start
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
		use std::cmp::{min, max};

		let mut dp = 0;
		let mut buy = prices[0];

		for i in prices
		{
			buy = min(buy, i);
			dp = max(dp, i - buy);
		}

		return dp;
    }
}
// @lc code=end
/*
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
		use std::cmp::max;

		let mut diff = 0;

		for i in 0..prices.len()
		{
			for j in i + 1..prices.len()
			{
				diff = max(diff, prices[j] - prices[i]);
			}
		}
		return diff;
    }
}
*/

File: python_src/121.买卖股票的最佳时机.py
#
# @lc app=leetcode.cn id=121 lang=python3
#
# [121] 买卖股票的最佳时机
#
from typing import *

# @lc code=start
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        ret = 0
        buy = prices[0]

        for price in prices[1:]:
            if price < buy:
                buy = price
            else:
                tmp = price - buy
                if tmp > ret: ret = tmp
        return ret
# @lc code=end

s = Solution()
print(s.maxProfit([7,1,5,3,6,4]))
print(s.maxProfit([7,6,4,3,1]))
